### ISSUE TITLE
Missing "url" property for DataSourceProperties will cause NullPointerException if embedded DB is not used

### ISSUE THREAD

**Issue by simonwg** [spring-projects/spring-boot#8690]
The issued failed with Spring Boot v1.5.2.

In my case, I am trying to create 2 `DataSource` (not any Embedded Database), Some configuration settings are common (e.g. Connection Pool type,  Driver name) and specific configuration (such as url, username, password etc) will be located in different configuration keys.

e.g.
```
application.yml

app:
    datasource:
        type: com.zaxxer.hikari.HikariDataSource
        driverClassName: com.mysql.jdbc.Driver
        db1:
            url: jdbc:mysql://localhost:3306/myDB1
            username: user1
            password: password1
        db2:
            url: jdbc:mysql://localhost:3306/myDB2
            username: user2
            password: password2
```

```
@Configuration
public class RepositoryConfig {

    @Bean
    @Primary
    @ConfigurationProperties("app.datasource")
    public DataSourceProperties dataSourceProperties() {
        return new DataSourceProperties();
    }

    @Bean
    @Primary
    @ConfigurationProperties("app.datasource.db1")
    public HikariDataSource db1DataSource(DataSourceProperties properties) {
        return (HikariDataSource) dataSourceProperties().initializeDataSourceBuilder()
                .type(HikariDataSource.class).build();
    }

    @Bean
    @ConfigurationProperties("app.datasource.db2")
    public HikariDataSource db2DataSource(DataSourceProperties properties) {
        return (HikariDataSource) dataSourceProperties().initializeDataSourceBuilder()
                .type(HikariDataSource.class).build();
    }
}
```

When I start with the above configuration, `NullPointerException` will be thrown.

```
Caused by: java.lang.NullPointerException: null
	at java.util.regex.Matcher.getTextLength(Matcher.java:1283) ~[na:1.8.0_121]
	at java.util.regex.Matcher.reset(Matcher.java:309) ~[na:1.8.0_121]
	at java.util.regex.Matcher.<init>(Matcher.java:229) ~[na:1.8.0_121]
	at java.util.regex.Pattern.matcher(Pattern.java:1093) ~[na:1.8.0_121]
	at java.util.Formatter.parse(Formatter.java:2547) ~[na:1.8.0_121]
	at java.util.Formatter.format(Formatter.java:2501) ~[na:1.8.0_121]
	at java.util.Formatter.format(Formatter.java:2455) ~[na:1.8.0_121]
	at java.lang.String.format(String.java:2940) ~[na:1.8.0_121]
	at org.springframework.boot.autoconfigure.jdbc.EmbeddedDatabaseConnection.getUrl(EmbeddedDatabaseConnection.java:109) ~[spring-boot-autoconfigure-1.5.2.RELEASE.jar:1.5.2.RELEASE]
	at org.springframework.boot.autoconfigure.jdbc.DataSourceProperties.determineUrl(DataSourceProperties.java:288) ~[spring-boot-autoconfigure-1.5.2.RELEASE.jar:1.5.2.RELEASE]
	at org.springframework.boot.autoconfigure.jdbc.DataSourceProperties.initializeDataSourceBuilder(DataSourceProperties.java:183) ~[spring-boot-autoconfigure-1.5.2.RELEASE.jar:1.5.2.RELEASE]

```

My current workaround is add a dummy url and let is overridden by specific property for each DB.
`app.datasource.url = dummy`

Please note that there is no embedded DB library for dependency in my application, otherwise everything works fine.

**Comment by wilkinsona** [comment]
The diagnostics could certainly be better, but the root cause here appears to be that you're using Boot's `DataSourceProperties` with your own, differently structured configuration. I would recommend that you don't do that and use your own properties class instead.

**Comment by snicoll** [comment]
Indeed. You are using a public method of `DataSourceProperties` that is meant to be used against several properties and you've relocated some of those somewhere else. In your case, not having a `url` there triggers a lookup for an embedded database (which doesn't make sense since you have well-defined urls).

**Comment by snicoll** [comment]
Let's fix the error message though

**Comment by snicoll** [comment]
The error message is fixed as part of #10626
